source/distributed/shared_tria.cc

```diff
@@ -134,6 +135,19 @@ namespace parallel
       this->update_number_cache ();
     }
 
+    template <int dim, int spacedim>
+    void
+    Triangulation<dim, spacedim>::
+    copy_triangulation (const dealii::Triangulation<dim, spacedim> &other_tria)
+    {
+      Assert ((dynamic_cast<const dealii::parallel::distributed::Triangulation<dim,spacedim> *>(&other_tria) == NULL),
```

Comment:
you are right, it is a tricky issue. I am fine with the current code unless @bangerth or @tjhei have suggestions on alternative implementation of the Assert which fails with `p::d::Tria` and pass with `p::s::Tria` and `::Tria`.